blockprod/src/detail/mod.rs

```diff
@@ -47,158 +67,516 @@ impl BlockProduction {
+                );
+                return Err(BlockProductionError::TipChanged(
+                    tip_at_start.block_id(),
+                    tip_at_start.block_height(),
+                    current_tip_index.block_id(),
+                    current_tip_index.block_height(),
+                ));
+            }
+
+            // TODO: instead of the following static value, look at
+            // self.chain_config for the current block reward, then send
+            // it to self.reward_destination
+            let block_reward = BlockReward::new(vec![]);
+
+            // TODO: see if we can simplify this
+            let transactions = match transactions_source.clone() {
+                TransactionsSource::Mempool => {
+                    self.collect_transactions().await?.transactions().clone()
+                }
+                TransactionsSource::Provided(txs) => txs,
+            };
+
+            let block_body = BlockBody::new(block_reward, transactions);
+
+            let tx_merkle_root = calculate_tx_merkle_root(&block_body)
+                .map_err(BlockCreationError::MerkleTreeError)?;
+            let witness_merkle_root = calculate_witness_merkle_root(&block_body)
+                .map_err(BlockCreationError::MerkleTreeError)?;
+
+            let block_header = BlockHeader::new(
+                current_tip_index.block_id(),
+                tx_merkle_root,
+                witness_merkle_root,
+                timestamp,
+                consensus_data,
+            );
+
+            // A synchronous channel that sends only when the mining/staking is done
+            let (ended_sender, ended_receiver) = mpsc::channel::<()>();
+
+            // Return the result of mining
+            let (result_sender, mut result_receiver) = oneshot::channel();
+
+            {
+                let chain_config = Arc::clone(&self.chain_config);
+                let current_tip_height = current_tip_index.block_height();
+                let stop_flag = Arc::clone(&stop_flag);
+
+                self.mining_thread_pool.spawn(move || {
+                    let block_header = Self::solve_block(
+                        chain_config,
+                        block_header,
+                        current_tip_height,
+                        stop_flag,
+                    );
+
+                    // This can fail if the function exited before the mining thread finished
+                    let _send_whether_ended = ended_sender.send(());
+
+                    result_sender
+                        .send(block_header)
+                        .expect("Failed to send block header back to main thread");
+                });
+            }
+
+            tokio::select! {
+                _ = cancel_receiver.recv() => {
+                    stop_flag.store(true, std::sync::atomic::Ordering::SeqCst);
+
+                    // This can fail if the mining thread has already finished
+                    let _ended = ended_receiver.recv();
```

Comment:
I was concered what would happen if we never reach send because of some error in the spawned thread because this `recv()` blocks, but luckily if there's a disconnect then `recv()` returns an error right away and so we won't end up blocking forever.

Reply:
Done